File: typescript/src/llmcontext/manager.ts
import type { Message } from "../llm/index.js";

export interface ToolResultEntry {
  call_id: string;
  tool_name: string;
  args: Record<string, unknown>;
  output: string;
  success: boolean;
}

export interface ContextManager {
  appendPrompt(jobId: string, msg: Message): void;
  appendResponse(jobId: string, msg: Message): void;
  appendToolCalls(jobId: string, calls: unknown[]): void;
  appendToolResults(jobId: string, results: ToolResultEntry[]): void;
  getHistory(jobId: string): Message[];
  purge(jobId: string): void;
}

export class InMemoryContextManager implements ContextManager {
  private history: Map<string, Message[]> = new Map();

  appendPrompt(jobId: string, msg: Message): void {
    this.ensureHistory(jobId).push(msg);
  }

  appendResponse(jobId: string, msg: Message): void {
    this.ensureHistory(jobId).push(msg);
  }

  appendToolCalls(_jobId: string, _calls: unknown[]): void {
    // Not stored in this implementation
  }

  appendToolResults(_jobId: string, _results: ToolResultEntry[]): void {
    // Not stored in this implementation
  }

  getHistory(jobId: string): Message[] {
    return this.history.get(jobId) || [];
  }

  purge(jobId: string): void {
    this.history.delete(jobId);
  }

  private ensureHistory(jobId: string): Message[] {
    if (!this.history.has(jobId)) {
      this.history.set(jobId, []);
    }
    return this.history.get(jobId)!;
  }
}

File: go/llmcontext/manager.go
package llmcontext

import (
	"github.com/soypete/pedro-agentware/go/llm"
	"github.com/soypete/pedro-agentware/go/toolformat"
)

// ContextManager manages the durable conversation history for an agent run.
// The default implementation writes each entry to a file, making the run
// recoverable across process restarts.
type ContextManager interface {
	// AppendPrompt records an outbound prompt message.
	AppendPrompt(jobID string, msg llm.Message) error
	// AppendResponse records an inbound LLM response.
	AppendResponse(jobID string, msg llm.Message) error
	// AppendToolCalls records the parsed tool calls for this round.
	AppendToolCalls(jobID string, calls []toolformat.ParsedToolCall) error
	// AppendToolResults records the results of tool executions for this round.
	AppendToolResults(jobID string, results []ToolResultEntry) error
	// GetHistory reconstructs the full message history for a job.
	GetHistory(jobID string) ([]llm.Message, error)
	// Purge deletes all context for a job (call after completion in production).
	Purge(jobID string) error
}

// ToolResultEntry pairs a tool call with its result for context storage.
type ToolResultEntry struct {
	CallID   string
	ToolName string
	Args     map[string]any
	Output   string
	Success  bool
}
